src/template-expr.adoc

```diff
@@ -0,0 +1,255 @@
+
+[{nrm}]
+----
+(*macro* decimal_constraint
+    [(precision *int{asterisk}*), (exponent *int{asterisk}*)]
+    {
+        precision: (*if_many* precision range::[precision] precision),
+        exponent:  (*if_many* exponent  range::[exponent]  exponent),
+    })
+----
+
+[{nrm}]
+----
+(:decimal_constraint (3) (-1))     ⇒ { precision: 3, exponent: -1 }
+(:decimal_constraint (1 5) (-5 0)) ⇒ { precision: range::[1, 5],
+                                       exponent: range::[-5, 0] }
+(:decimal_constraint (:) (3 max))  ⇒ { exponent: range::[3, max] }
+(:decimal_constraint (1) (:))      ⇒ { precision: 1 }
+----
+
+
+==== Mapping
+
+These special forms produce repeated output mapped across elements of a stream.
+
+===== `*for*`
+
+[{nrm}]
+----
+(*for* [(_id_ _in-template_), ...] _body-template_)
+----
+
+Iteratively expands the _body-template_ using individual values from the _in-templates_.
+
+Each iteration takes the next value from each _in-template_ stream; iteration stops when any
+stream ends.
+Local variables are created for each identifier _id_, bound to the current value from their stream.
+The _body-template_ is then expanded in that environment, and iteration proceeds.
+The result of the `*for*` expression is the concatenated results of the body expansions.
+
+NOTE: The termination rule is under discussion; see
+https://github.com/amazon-ion/ion-docs/issues/201
+
+
+=== Macro Invocation
+
+A macro definition can express its output in terms of other macros. Quite often, these will be
+macros provided by the Ion implementation,
+// —everything in System Macros 2023-05 is available by default—
+but they can also be acquired from other modules.
+
+// TODO link to system-macro chapter
+
+The S-expression syntax for macro invocation is similar to that of E-expressions.
+When a template is an S-expression and the first element is not the name of a special form, that
+element must instead be a _macro-ref_ and the template denotes a macro invocation.
+There are multiple sources of macros: the defining module’s internal environment (which is being
+incrementally extended with each definition), and the exported macros of modules loaded by the
+enclosing module or `$ion_encoding` directive.
+
+// TODO See Resolving Macro References: Encoding Modules 2023-05 for the relevant algorithm.
+
+The remaining elements of the S-expression are subforms that denote the inputs to the macro.
+These use normal Ion notation, but what’s syntactically acceptable is defined by the macro’s
+signature.
+
+The number of such subforms (that is, the invocation’s actual arity) must be equal to or greater
+than the macro’s minimum arity, and at most its maximum arity, when one exists.
+In other words, an invocation must contain one subform for each required parameter, followed by
+optional subforms for the remaining optional parameters.
+
+Within an invocation expression, the syntax of each subform is defined first by its parameter’s
+grouping form, then its base type:
```

Comment:
I think some examples would make this section easier for folks to digest without needing to refer to the glossary. (I'm fine with a `TODO`.)

Reply:
Added TODO